doc/guides/kconfig/tips.rst

```diff
@@ -416,6 +420,40 @@ interface, to make sure that things behave the way you expect. This is
 especially true when making moderately complex changes like these.
 
 
+Assignments to promptless symbols in configuration files
+********************************************************
+
+Assignments to invisible (promptless) symbols in configuration files are always
+ignored. Invisible symbols get their value indirectly from other symbols, via
+e.g. ``default`` and ``select``.
+
+A common source of confusion is opening the output configuration file
+(:file:`zephyr/.config`), seeing a bunch of assignments to invisible symbols,
+and assuming that those assignments must be respected when the configuration is
+read back in by Kconfig. In reality, all assignments to invisible symbols in
+:file:`zephyr/.config` are ignored by Kconfig, like for other configuration
+files.
+
+To understand why :file:`zephyr/.config` still includes assignments to
+invisible symbols, it helps to realize that :file:`zephyr/.config` serves two
+separate purposes:
+
+1. It holds the saved configuration, and
+
+2. it holds configuration output. :file:`zephyr/.config` is parsed by the CMake
+   files to let them query configuration settings, for example.
+
```

Comment:
I prefer *visible* and *invisible*, because it's the terminology you'll see in other places that talk about Kconfig. There's also *symbol visibility*.

I added another explanation of what *visible* means at the top of the page in this commit btw. It mentions that promptless symbols are always invisible, and that symbols with prompts can be invisible too.

> Really not a fan of the word "invisible" because it's invisible only in the interactive interfaces that some people do not use.

Yeah... I'm trying to encourage more people to use it. I know I keep nagging about it, but if people work with Kconfig without ever using the configuration interfaces, they're guaranteed to misunderstand a lot of things.

`menuconfig` should be front and center of people's minds, because it shows how Kconfig thinks. Enabling/disabling a symbol in a configuration file is just like toggling it on/off in `menuconfig`. If the symbol isn't visible, you can't toggle it on/off.

It's kinda inverted now, where people think `menuconfig` is this obscure thing that can safely be ignored most of the time, even though it's enormously important to get an understanding of how Kconfig works. :/

Reply:
Yes, having used menuconfig  is necessary to understand how Kconfig works. You can even stop at menuconfig if you're configuring your single, hobby system at home. And then you don't even know that some symbols are invisible because... you don't see them.

On the other hand, menuconfig is nowhere near enough to  "scale up" to managing fine grained policies across various hardware, software versions and concurrent developers. For that you need text files in source control as usual:
https://www.yoctoproject.org/docs/2.5/kernel-dev/kernel-dev.html#creating-config-fragments
https://www.chromium.org/chromium-os/how-tos-and-troubleshooting/kernel-configuration
https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/271688
etc.

The serious usability issue that this documentation PR is trying to mitigate is that  "visible" and "invisible"... *look* exactly the same! Hilarious.

> I prefer visible and invisible, because it's the terminology you'll see in other places that talk about Kconfig. There's also symbol visibility.

A good way to deal with terminology that bad in documentation is to treat as it were "deprecated": mentioned at the start of the section and then replaced in the rest.

"_We've always done it this way_"

Reply:
Just saying that you should be running `menuconfig` after every non-trivial change you make, to check things. I'd probably mess up a lot of things of I didn't.

No such thing as something being broken just in `menuconfig` by the way, which I think is a common misunderstanding (not saying you have it though). It's just that issues are much easier to spot there.

I think visible and invisible is good terminology. Shows up in lots of other places too, so inventing our own term might make things confusing.

Reply:
See "hidden" a lot too. I could sneak that in there if you want.